core/src/testUtils/groovy/nu/yona/server/YonaServer.groovy

```diff
@@ -172,4 +182,64 @@ class YonaServer
 	{
 		DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZ").format(dateTime)
 	}
+
+	static def relativeDateTimeStringToZonedDateTime(relativeDateTimeString)
+	{
+		def fields = relativeDateTimeString.tokenize(' ')
+		assert fields.size() <= 3
+		assert fields.size() > 0
+		int parsedFields = 0
+		int weekOffset = 0
+		int dayOffset = 0
+
+		LocalDateTime now = LocalDateTime.now()
+		switch (fields.size())
+		{
+			case 3:
+				assert fields[0].startsWith("W")
+				weekOffset = Integer.parseInt(fields[0].substring(1))
+				parsedFields++
+			// Fall through
+			case 2:
+				int weekDay = DateTimeFormatter.ofPattern("eee").parse(fields[parsedFields]).get(ChronoField.DAY_OF_WEEK)
+				dayOffset = weekDay - now.dayOfWeek.value
+				parsedFields++
+			// Fall through
+			case 1:
+				ZonedDateTime dateTime = parseTimeForDay(fields[parsedFields], now.plusDays(dayOffset).plusWeeks(weekOffset).getLong(ChronoField.EPOCH_DAY))
+				assert dateTime.compareTo(ZonedDateTime.now()) <= 0 // Must be in the past
+				return dateTime
+		}
+	}
+
+	private static ZonedDateTime parseTimeForDay(String timeString, long epochDay)
+	{
+		DateTimeFormatter formatter =
+				new DateTimeFormatterBuilder().appendPattern("HH:mm[:ss][.SSS]")
+				.parseDefaulting(ChronoField.EPOCH_DAY, epochDay)
+				.toFormatter()
+				.withZone(ZoneId.of("Europe/Amsterdam"))
+		ZonedDateTime.parse(timeString, formatter)
+	}
+
+	static def relativeDateStringToDaysOffset(weeksBack, shortDay)
+	{
+		int weekDay = getDayOfWeek(DateTimeFormatter.ofPattern("eee").parse(shortDay).get(ChronoField.DAY_OF_WEEK))
+		return weeksBack * 7 - weekDay + LocalDateTime.now().dayOfWeek.value
```

Comment:
Groovy is Java related but different. I prefer to leverage the simplifications that Groovy offers over Java. We could do many things Java-style that we now do Groovy-style (think of loops for instance).